test/PackageVersionsTest/FallbackVersionsTest.php

```diff
@@ -54,9 +54,15 @@ public function testValidVersions() : void
 
     public function testValidVersionsWithoutComposerLock() : void
     {
+        // This test will always fail if there is nothing in the installed.json
+        // due to this package being installed with `composer install --no-dev`.
+        if (json_decode(file_get_contents(getcwd() . '/vendor/composer/installed.json', true)) === []) {
+            $this->markTestSkipped('Empty installed.json (possible --no-dev)');
```

Comment:
We should have a new test for `--no-dev`, not modifying the existing ones

Reply:
The problem is that when the `--no-dev` flag is given, this test will always fail. This logic is not to test the validity of `--no-dev`, but just to say this test is not useful because of it. How should I address this failing test when `--no-dev` was used?

Reply:
This part of the test suite is supposed to ve run with dev dependencies anyway, since we don't have phpunit at all otherwise.

This conditional can most likely be dropped, since the unit tests will never run under those conditions. Different story for e2e tests

Reply:
Got it. I just checked the travis config, and there is no e2e test specifying `--no-dev`, so as of right meow, they should still pass. 

As for adding a new e2e test specifically for this, I'd recommend adding `@group` annotations to the tests so that this test can be excluded from the `--no-dev` tests without having to add logic here.

Reply:
Yeah, `--no-dev` is not really testable unless we write a script that works without any dependencies at all. Since `--no-dev` is the de-facto way of installing in production, when not having a located `composer.lock`, I suppose the only side effect is not having versions for the dev dependencies, right?

Reply:
Maybe instead of getting the data from the `composer.lock`, we get it from `installed.json`. Because `composer.lock` will always have the dev dependencies, any time `--no-dev` is passed this will always fail. But, since we're validating that this method should fallback to `installed.json`, it makes sense (to me) to test against its data. So, in this case, it should merely skip the meat of the test automatically, assuming we remove the assertion on the empty `$packages` array.

Your call though.

Reply:
I'd say that the `composer.lock` is correct here, even if the packages aren't installed.

We can roll with the patch as-is, if you are happy with it.